### PR TITLE
Implement work-stealing scheduler

### LINKED ISSUE #858: Implement work-stealing scheduler
The more I look at the scheduling problem, the more I think it should be solved using work stealing https://en.wikipedia.org/wiki/Work_stealing :

- Initially, all tests are split among the workers evenly.
- When some worker completes all of its tests, the scheduler asks the worker with max number of pending tests to give half of its pending tests back.
- After receiving the confirmation, the scheduler sends the returned tests to the worker that ran out of tests.
- Shutdown workers when no tests can be reallocated ("stolen").

Of course, there are some tricky synchronization details - but nothing impossible to implement. This algorithm shouldn't need any parameters or assumptions about test duration to perform optimally. And the code could even turn out to be simpler than the current `LoadScheduling`.

By "perform optimally" I mean:
- In the ideal case, where all test have the same duration: `n` tests are split among `m` workers, each worker runs `n/m` consecutive tests - and achieves the best reuse of fixtures possible (possible without additional information about the fixtures).
- No workers are idle while there are pending tests.

The only issue is that I don't know when I'll have enough free time to write a custom scheduler from scratch.

I plan to work on this myself, but I'm not sure when I'll have time for it.

Also, I would appreciate feedback on the idea from the maintainers.

### PR DESCRIPTION
**Pull request by amezin**
Closes #858

### CODE REVIEW

**Comment by nicoddemus** [comment]
Thanks @amezin!

I will review it in the next few days.

You have made several contributions already, would you like to help maintain the project and become a contributor?

**Comment by amezin** [comment]
What does "becoming a contributor" mean? GitHub already shows the "Contributor" badge...

**Comment by nicoddemus** [comment]
Basically write access, being able to review/merge PRs, and making releases.

File: src/xdist/scheduler/worksteal.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+        self.steal_requested = None
+
+        indices_set = set(indices)
+        self.node2pending[node] = [
+            i for i in self.node2pending[node] if i not in indices_set
+        ]
+        self.pending.extend(indices)
+        self.check_schedule()
+
+    def check_schedule(self):
```
Comment (nicoddemus) [inline src/xdist/scheduler/worksteal.py:174]: Great comments along the code. 💯 👍

**Review by nicoddemus — COMMENTED** [review]
This looks really good!

Left a few minor comments which would be nice to get in as well.

Did you test this new mode in a large test suite?

**Comment by nicoddemus** [comment]
Feel free to use type-annotations as well, as we run mypy on the repository. 👍

File: src/xdist/dsession.py [hunk 2]
```diff
@@ -282,6 +284,17 @@ def worker_runtest_protocol_complete(self, node, item_index, duration):
         """
         self.sched.mark_test_complete(node, item_index, duration)
 
+    def worker_unscheduled(self, node, indices):
+        """
+        Emitted when a node fires the 'unscheduled' event, signalling that
+        some tests have been removed from the worker's queue and should be
+        sent to some worker again.
```
[the remark this answers is not in the document]
Comment (amezin) [reply src/xdist/dsession.py:291]: Actually, there's no requirement that these tests shouldn't be sent to the same worker again. Yes, usually it's pointless... But, for example, if all other workers were removed between "steal" request and "unscheduled" response, there's no other option than to send all tests back to the same worker.

File: src/xdist/scheduler/worksteal.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+
+NodePending = namedtuple("NodePending", ["node", "pending"])
+
+# Every worker needs at least 2 tests in queue - the current and the next one.
+MIN_PENDING = 2
+
+
+class WorkStealingScheduling:
+    """Implement work-stealing scheduling.
```
[the remark this answers is not in the document]
Comment (amezin) [reply src/xdist/scheduler/worksteal.py:17]: Mostly copy-pasted from LoadScheduling

File: src/xdist/remote.py [hunk 4]
```diff
@@ -92,38 +97,54 @@ def pytest_sessionfinish(self, exitstatus):
...
+            for i in range(len(self.session.items)):
+                self.torun.put(i)
+        elif name == "shutdown":
+            self.torun.put(None)
+        elif name == "steal":
+            old_queue, self.torun = self.torun, self._make_queue()
+            stolen = []
+            steal_set = set(kwargs["indices"])
+            try:
+                while 1:
```
[the remark this answers is not in the document]
Comment (amezin) [reply src/xdist/remote.py]: After looking at these lines again, I've changed the infinite loop to a loop over `iter(queue.get_nowait, ...)` (well, a bit more complicated) instead. I think it should be a bit more readable (but I'm not 100% sure)

**Comment by amezin** [comment]
> Did you test this new mode in a large test suite?

https://github.com/ddterm/gnome-shell-extension-ddterm/actions/runs/3843994508/jobs/6546739410

Speed/time seems to be indistinguishable from `load` - either when the suite is split to multiple jobs, or when I run all tests at once (~4k tests total). On GitHub CI, run to run variance is higher than `load` vs `worksteal` difference. When I run the same suite locally, I get maybe 5% speedup compared to `load`, but nothing really noticeable.

**Comment by amezin** [comment]
> Basically write access, being able to review/merge PRs, and making releases.

Yes, I'm interested. But can't promise I'll be consistently active.

**Comment by nicoddemus** [comment]
> Yes, I'm interested. But can't promise I'll be consistently active.

Sure, no problem, we have no such expectations. :+1:

**Review by nicoddemus — APPROVED** [review]
Thanks @amezin!

Will merge this in the next few days to give the chance for others to review too. :+1:

File: src/xdist/dsession.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nicoddemus) [reply src/xdist/dsession.py:291]: Fair enough!

**Pull request merged**